### PR TITLE
If we read in bad data from disk, do not blindly use it as it may not  valid wrt the state of the project.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide .  I believe this may be a fix for https://devdiv.visualstudio.com/DevDiv/_workitems?id=359238&triage=true&_a=edit

File: src/Workspaces/Core/Portable/FindSymbols/DeclaredSymbolInfo.cs [hunk 1]
```diff
@@ -79,48 +81,52 @@ internal void WriteTo(ObjectWriter writer)
...
+                var node = root.FindNode(this.Span);
+                return semanticModel.GetDeclaredSymbol(node, cancellationToken);
+            }
+            else
+            {
+                var message =
+$@"Invalid span in {nameof(DeclaredSymbolInfo)}.
+{nameof(this.Span)} = {this.Span}
+{nameof(root.FullSpan)} = {root.FullSpan}";
+                FatalError.ReportWithoutCrash(new InvalidOperationException(message));
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/DeclaredSymbolInfo.cs]: Do you think this gives you enough information to fix the issue from the reports?  Just knowing the two different spans?

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/DeclaredSymbolInfo.cs]: It's at least better than today where i have no information.  I'd like to at least ideally know if, say, we have some sort of 'empty' span in one of these.  perhaps indicating that we could not even read in a document properly.

File: src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs [hunk 2]
```diff
@@ -18,16 +18,16 @@ internal class DeclaredSymbolNavigableItem : INavigableItem
             public ImmutableArray<TaggedText> DisplayTaggedParts => _lazyDisplayTaggedParts.Value; 
 
             public Document Document { get; }
-            public Glyph Glyph => Symbol?.GetGlyph() ?? Glyph.Error;
+            public Glyph Glyph => this.SymbolOpt?.GetGlyph() ?? Glyph.Error;
             public TextSpan SourceSpan => _declaredSymbolInfo.Span;
-            public ISymbol Symbol => _lazySymbol.Value;
+            public ISymbol SymbolOpt => _lazySymbolOpt.Value;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs]: Correct.

**Comment by CyrusNajmabadi** [comment]
> Is the thought is that this will address the issue because we won't used the cached data, we'll re-create it from a current snapshot?

Yes.  And i've also gone through and ensured that if some bad data leaks out, we at least report it and are resilient to it in case we ahve some sort of 'out-of-sync' issue.

Note: out-of-sync should be impossible (as we compare syntax versions).  But i'm not ignoring the possibility that such a thing may happen due to an underlying workspace bug.

**Comment by Pilchie** [comment]
@dotnet/roslyn-ide Can we get a second review?

Tagging @MattGertz for potential RC.3 inclusion.

**Comment by CyrusNajmabadi** [comment]
**Customer scenario**

Customer invokes navigate to.  IDE crashes.

**Bugs this fixes:** 

https://devdiv.visualstudio.com/DevDiv/_workitems?id=359238&triage=true&_a=edit

**Workarounds, if any**

Customer does not perform navigate to.

**Risk**

Low.

**Performance impact**

Low.  All we're doing is checking any being resilient to having bad data.

**Is this a regression from a previous update?**

Unclear.

**Root cause analysis:**

The root cause is currently unclear.  The system is designed so that what's going on shoudl be impossible.  The changes here make it so we're at least resilient to it happening, and so we can collect more data to help track down what the core issue is.

**How was the bug found?**

Watsons

**Pull request merged**